server/test_nlp_manual.py

```diff
@@ -31,25 +39,162 @@ def _encoded_q(query: str, embedding: str = 'us_filtered'):
   return urllib.parse.urlencode({'q': query, 'build': embedding})
 
 
-def test_palo_alto_flow():
-  query = "tell me about Palo Alto"
-  response = app.test_client().post(f'/nl/data?{_encoded_q(query)}',
-                                    json={'contextHistory': []},
-                                    content_type='application/json')
+# The following are compact representations of nl API result.
+# Feel free to add what is needed in order to verify the test flow.
+
+
+@dataclass
+class Tile:
+  stat_var_key: Optional[str] = None
+  ranking: Optional[str] = None
+
+
+@dataclass
+class Categories:
+  # Checks categories[0]['blocks'][0]['columns'][0]['tiles']
+  # Order is not tested.
+  tiles: Optional[List[Tile]] = None
+
+
+@dataclass
+class Metadata:
+  place_dcid: Optional[List[str]] = None
+
+
+@dataclass
+class Config:
+  categories: Optional[Categories] = None
+  metadata: Optional[Metadata] = None
+
+
+def _verify_categories(categories: Categories, got: Dict):
+  if categories.tiles:
+    # Check if all tile specs in this this has stat var keys
+    assert all(t.stat_var_key for t in categories.tiles)
+
+    stat_var_keys_got = set()
+    for tile in got[0]['blocks'][0]['columns'][0]['tiles']:
```

Comment:
since this is being invoked a couple of times, could this be in a helper function? This may also end up being fragile due to the the fairly complex path to get tiles

Reply:
oops, I forgot I had auto merge on. Will iterate on improving the logic here. Completely agree that this is very fragile.